Create exclude-patterns.md

benwuNYC opened the pull request:
**Fixes issue #**: #355 (--exclude doesn't exclude when the pattern is an absolute path)

**Description of the changes being introduced by the pull request**:
Comprehensive documentation on --exclude that describes special characters and provides some examples, points to existing documentation of .gitignore and Pathspec. 

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [X] Tests have been added for the bug fix or new feature
- [X] Docs have been added for the bug fix or new feature

benwuNYC committed:
Create exclude-patterns.md

Documentation for the --exclude command-line argument for in-toto run.

Signed-off-by: Benjamin Wu <benjaminwu1337@gmail.com>

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,93 @@
+- Pattern Format
+  - Include backslash (' \\ ') in front of first hash for patterns that begin with a literal hash
+  - Trailing spaces are ignored unless quote with backslash (' \\ ')
+  - Prefix '!' negates the pattern. Excluded patterns will become included again 
```
> Excluded patterns will become included again

I think this can be a little confusing. Let's talk about it in terms of the files matched by the pattern as a whole.

benwuNYC replied:
Removed the comment from pattern formats as a whole and think it would be more appropriate in the special patterns section and changed up the wording.

adityasaky commented on doc/source/exclude_patterns.md:
````diff
@@ -0,0 +1,93 @@
+# Exclude Pattern Settings 
+- Optional argument for [in-toto-run.](https://in-toto.readthedocs.io/en/latest/command-line-tools/in-toto-run.html) Path patterns to match paths that should not be recorded as 'materials' or 'products'.  Passed patterns override patterns defined in environment variables or config files (see ```ARTIFACT_EXCLUDE_PATTERNS``` documentation for additional info) 
````
Line length alert!

Let's keep them <= 80. Also note that you don't need to add a space at the end of a line or beginning of the next line.

benwuNYC replied:
Fixed it up a bit, let me know if the length works!

adityasaky commented on doc/source/exclude_patterns.md:
````diff
@@ -0,0 +1,93 @@
+# Exclude Pattern Settings 
+- Optional argument for [in-toto-run.](https://in-toto.readthedocs.io/en/latest/command-line-tools/in-toto-run.html) Path patterns to match paths that should not be recorded as 'materials' or 'products'.  Passed patterns override patterns defined in environment variables or config files (see ```ARTIFACT_EXCLUDE_PATTERNS``` documentation for additional info) 
+- Uses [PathSpec](http://python-path-specification.readthedocs.io) to compile [gitignore-style patterns](https://git-scm.com/docs/gitignore), implementing GitWildMatchPattern class which itself is derived from [Rsync's wildmatch](https://github.com/AndyA/rsync/blob/master/lib/wildmatch.c)   
````
I think this should be clarified. Do you mean git's patterns are derived from rsync's wildmatch?

benwuNYC replied:
Yep! I think that we might not need that detail since we're focusing on and linking the Pathspec documentation, which already does a fantastic job describing how Git's patterns are derived from RSync's wild match.

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,93 @@
+  - Include backslash (' \\ ') in front of first hash for patterns that begin with a literal hash
+  - Trailing spaces are ignored unless quote with backslash (' \\ ')
+  - Prefix '!' negates the pattern. Excluded patterns will become included again 
+  - The slash (' / ') is used as a directory separator. Separators at beginning or middle (or both) of pattern are relative to directory level of particular .gitignore file. If there is a separator at the end of the pattern then the pattern will only match directories. 
```
I wonder if backticks are appropriate here (and elsewhere).

```suggestion
  - The slash (`/`) is used as a directory separator. Separators at beginning or middle (or both) of pattern are relative to directory level of particular .gitignore file. If there is a separator at the end of the pattern then the pattern will only match directories. 
```

benwuNYC replied:
Decided to remove all the said characters and only include it in text, except in examples!

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,93 @@
+  - Pattern ending with double-asterisks will match any trailing path segments
+  - Pattern with inner double-asterisks will match multiple (or zero) inner path segments 
+
+# Special Characters / Edge Cases 
```
```suggestion
## Special Characters / Edge Cases 
```

Sub headings :)

benwuNYC replied:
Fixed, thanks for the heads up :) . Also decided to rename it to "Special Patterns" as it seems more fitting.

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,93 @@
+  - Pattern with inner double-asterisks will match multiple (or zero) inner path segments 
+
+# Special Characters / Edge Cases 
+1. **Patterns beginning with ' / '**
```
```suggestion
### Patterns beginning with ' / '
```
sub sub heading?

benwuNYC replied:
I was thinking that might be slightly confusing since it looks similar the sub-heading, but it still looks organized. Updated them, but let me know if you share the same concerns :)

adityasaky commented on doc/source/exclude_patterns.md:
````diff
@@ -0,0 +1,93 @@
+  - The pattern ```doc/frotz``` and ```/doc/frotz``` have the same effect as a ```.gitignore file```. In other words, a leading slash is not relevant if there is already a middle slash in the pattern.
+  - The pattern ```foo/*```, matches ```foo/test.json``` (a regular file), ``` foo/bar ``` (a directory), but it does not match ``` foo/bar/hello.c``` (a regular file), as the asterisk in the pattern does not match ``` bar/hello.c```  which has a slash in it.
+
+# Source Code
````
I think we don't need the source code section. Perhaps we can include a link, but I'm still wary as we'd have to go by line number which can change. Perhaps just a mention to look in runlib.

benwuNYC replied:
Agreed, especially since it IS a helper function.

adityasaky commented on doc/source/exclude_patterns.md:
````diff
@@ -0,0 +1,93 @@
+ return sorted(included)
+ ``` 
+
+# Exclude Pattern Arguments
````
`_apply_exclude_patterns` is "private" so I'm not certain we need this in the doc.

benwuNYC replied:
Fixed!

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,93 @@
+
+
+
+# Tutorial
```
I wonder if the pattern examples above suffices. I also doubt we want people using `_apply_exclude_patterns` directly. It's a helper function. :)

benwuNYC replied:
Yep, removed the tutorials, source-code, and arguments section!

adityasaky reviewed (changes requested):
I think it's a great start at resolving situations like in #355, @itsbenwu, but I think it needs a bit more work. And thanks for moving this document to GitHub!

I've left some thoughts below, and I also wonder if we want to go bullet points all the way. I think the initial information could go in a paragraph or two and the bullet points can be reserved for "Pattern Format" and so on.

lukpueh reviewed (commented):
Thanks a ton for taking a stab at this, @itsbenwu, this is solid work! :)

Here is some high-level feedback that mostly aligns with @adityasaky's review.

The typical use case for this document is a user of the `in-toto-run` command line tool, or `runlib.in_toto_run` library function, wanting to know more about the exact format of the `--exclude` (cli) or `exclude_patterns` (api) argument (or `ARTIFACT_EXCLUDE_PATTERNS` setting). So here is how I'd structure this document:

----8<----

1. A paragraph that briefly explains materials and products in the context of generating link metadata with in-toto tools, and how exclude patterns fit in. This can be very abstract, you can talk about the in-toto tools and how to use them if you think it is helpful, but you don't have to. This document should really be first and foremost about the exclude patterns.

2. The pattern format specification: A list of bullet points, each describing a special character and how it is interpreteted.

3. Examples: Again, this can be very abstract. I wouldn't add code snippets of any tools or functions, but instead just strings of patterns and which files they would and wouldn't match.

4. (optional) Pointers to underlying technology, PathSpec, gitwildmatch, etc... and corresponding documentation.


----8<----

Most of these things are already there in your draft, so maybe you can just give it another pass, and re-structure and re-work things a little bit. Try to look at the document from the perspective of the in-toto user, who wants to quickly write patterns in order to exclude certain files from link metadata.

benwuNYC commented:
Thanks for the review and feedback @lukpueh and @adityasaky! Under your recommendations, I did the following

1. I included a brief paragraph explaining the materials and products in the context of generating link metadata and how the exclude patterns fit in. I didn't go too much in detail of in-toto-run as I'm assuming the user already knows the general details of in-toto-run, but let me know if you think a brief introduction or summary would be beneficial.

2. In the pattern format section, I included details on how patterns should be written and in the special patterns, section described each pattern and how it would be interpreted. 

3. Sure, I'll find some examples and see if they fit in this context. 

4. Done! Documentation to Pathspec and .gitignore all included in the documentation section. 

Overall, I've tried to cut it short and simple for any users to understand. Will have the PR ready VERY soon! :)

benwuNYC commented:
Thanks for the review @adityasaky! Under your recommendations, I did the following: 

- Formatted the beginning of the documentation into a brief paragraph, incorporating it in the context of generating link metadata. 
- Removed some unnecessary details (Rsync) to keep the document short and simple.
- Reformatted pattern formats to include only details on how patterns should be written, while the special pattern section details how each character should be interpreted. 
- Fixed up long lines.
- Removed backticks and unnecessary parenthesis. 
- Renamed a few sections
- Changed bolded statements to sub-sub headings. 
- Removed source code, tutorial, arguments sections.

May have to fix up a few more examples and see where I could cut it a bit shorter but I should and hopefully be close to finished! 🐳

adityasaky commented on doc/source/exclude_patterns.md:
````diff
@@ -0,0 +1,44 @@
+# Exclude Pattern Settings 
+```--exclude``` is an optional command-line argument for [in-toto-run](https://in-toto.readthedocs.io/en/latest/command-line-tools/in-toto-run.html) that uses [PathSpec](http://python-path-specification.readthedocs.io) while ```exclude_pattern``` is an API argument that follows [gitignore-style patterns](https://git-scm.com/docs/gitignore) and may be specified via the ```ARTIFACT_EXCLUDE_PATTERNS ...
````
> ```--exclude``` is an optional command-line argument for [in-toto-run](https://in-toto.readthedocs.io/en/latest/command-line-tools/in-toto-run.html)

So far so good.

> that uses [PathSpec](http://python-path-specification.readthedocs.io) while ```exclude_pattern``` is an API argument that follows [gitignore-style patterns](https://git-scm.com/docs/gitignore) and may be specified via the ```ARTIFACT_EXCLUDE_PATTERNS``` setting.

I think it gets confused here.

They aren't two separate things. We have [`runlib`](https://github.com/in-toto/in-toto/blob/develop/in_toto/runlib.py) which provides the `in_toto_run` API. This API accepts `exclude_patterns` as an optional argument. Then, we have have the `in-toto-run` CLI tool which accepts `--exclude-patterns` as an optional argument. However, they aren't doing anything separately. `in-toto-run` merely uses the API in `runlib` to provide all functionality, including exclude patterns. Any exclude patterns passed in via the CLI argument is just passed on to the API appropriately.

The API uses PathSpec to implement this functionality because we want gitignore-like patterns, which PathSpec promises.

benwuNYC replied:
Took a while to find the words to merge the API and CLI definitions into one, but I rewrote it below. 

The optional argument ```exclude_patterns``` in the ```in_toto_run``` API used by ```--exclude``` in the```in-toto-run``` command line tool applies [PathSpec](http://python-path-specification.readthedocs.io) to compile [gitignore-style patterns](https://git-scm.com/docs/gitignore).

adityasaky commented on doc/source/exclude_patterns.md:
````diff
@@ -0,0 +1,44 @@
+# Exclude Pattern Settings 
+```--exclude``` is an optional command-line argument for [in-toto-run](https://in-toto.readthedocs.io/en/latest/command-line-tools/in-toto-run.html) that uses [PathSpec](http://python-path-specification.readthedocs.io) while ```exclude_pattern``` is an API argument that follows [gitignore-style patterns](https://git-scm.com/docs/gitignore) and may be specified via the ```ARTIFACT_EXCLUDE_PATTERNS ...
+
+The feature prevents given paths and hashes of 'materials' (files before command execution) and 'products' (files after command execution) from being written together with other information (executed command, return value, stdout and stderr) in generating link metadata files. When a given pattern's directory is excluded, all its files and subdirectories are excluded as well.
````
While I think this is mostly right in spirit, we may want to talk about in terms of files matched by the pattern, and them being entirely excluded from being recorded as materials or products. Mentioning other information that we do record, while correct, may just confuse the reader.

> When a given pattern's directory is excluded, all its files and subdirectories are excluded as well.

Perhaps we can say "If a pattern matches a directory, all the files and subdirectories within that directory are also excluded."

benwuNYC replied:
Removed the details and reworded!

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,44 @@
+
+The feature prevents given paths and hashes of 'materials' (files before command execution) and 'products' (files after command execution) from being written together with other information (executed command, return value, stdout and stderr) in generating link metadata files. When a given pattern's directory is excluded, all its files and subdirectories are excluded as well.
+
+# Pattern Format
```
```suggestion
## Pattern Formats / Structures
```

adityasaky replied:
I'm not sure if just format is enough here, but I can really go either way

benwuNYC replied:
Format and structure seem to have the same definition, so it might be more simple to specify it as one or the other.

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,44 @@
+  - Question marks match any one character except a slash.
+
+## Special Patterns
+1. ### Pattern beginning with a slash ###
```
We don't need the numbers here, let's just use the `###` to subhead them.

Also, you don't need to "close" with a `###`.

The entire line can just be `### Patterns beginning with a slash`.

benwuNYC replied:
Agreed!

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,44 @@
+1. ### Pattern beginning with a slash ###
+  - Match paths on the root directly instead of any descendant paths by making a pattern relative to root. Escape the relative root conversion with an extra forward slash in front so as to match absolute paths. (i.e, "//"). 
+
+2. ### Pattern that is a slash ###
```
same here, and below

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,44 @@
+
+## Special Patterns
+1. ### Pattern beginning with a slash ###
+  - Match paths on the root directly instead of any descendant paths by making a pattern relative to root. Escape the relative root conversion with an extra forward slash in front so as to match absolute paths. (i.e, "//"). 
```
```suggestion
  - Match paths on the root directly instead of any descendant paths by making a pattern relative to root. Escape the relative root conversion with an extra forward slash in front so as to match absolute paths. (i.e., `//<rest of the pattern>`). 
```

I think we should use backticks when talking about pattern specific characters, as here.

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,44 @@
+  - Match paths on the root directly instead of any descendant paths by making a pattern relative to root. Escape the relative root conversion with an extra forward slash in front so as to match absolute paths. (i.e, "//"). 
+
+2. ### Pattern that is a slash ###
+  - According to 'git check-ignore' (v2.4.1), a single '/' does not match any file
```
I wonder how we want to do this. My distro, for example, has a newer version of git. Can we point to the current gitignore documentation for these (while also mentioning them)?

benwuNYC replied:
I've linked the current gitignore documentation in the documentation section, but it doesn't say anything about a single slash pattern. Furthermore, I can't seem to find the documentation for 'git check-ignore' (v2.4.1) to verify this but I included the 'git ignore' document the [Pathspec documentation](https://github.com/cpburnz/python-path-specification/blob/master/pathspec/patterns/gitwildmatch.py#L59-L61) refers to and utilizes.

adityasaky replied:
Yeah, I think we can remove references to `git check-ignore` which is a CLI tool to check against `.gitignore` for specific paths in the repository. Let's just state the fact and move on, and the links to the docs at the bottom should handle the rest, especially since we start the document naming both PathSpec and gitignore.

adityasaky commented on doc/source/exclude_patterns.md:
````diff
@@ -0,0 +1,44 @@
+  - Matched patterns will be unmatched. Escape the exclamation mark with a back-slash to match a literal exclamation mark (i.e, '\!').
+
+## Pattern Format Examples
+  - The pattern ``` hello.* ``` matches any file or folder whose name begins with **hello**. If one wants to restrict this only to the directory and not in its subdirectories, one can prepend the pattern with a slash, i.e. ``` /hello.*``` ; the pattern now matches ``` hello.txt, hello.c```  but not ``` a/hello.java ```. 
````
```suggestion
  - The pattern `hello.*` matches any file or folder whose name begins with **hello**. If one wants to restrict this only to the directory and not in its subdirectories, one can prepend the pattern with a slash, i.e. `/hello.*`; the pattern now matches `hello.txt, hello.c` but not `a/hello.java`. 
```

adityasaky replied:
There are some spacing inconsistencies here and elsewhere. Also, a single backtick is enough to enclose code in-line.

adityasaky replied:
Feel free to mark a review comment thread as resolved when done. :)

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,44 @@
+  - Beginning double asterisks match any leading path segments, inner double asterisks match multiple (or zero) inner path segments, and ending double asterisks match any trailing path segments. Single asterisks match everything except a slash. 
+  - Question marks match any one character except a slash.
+
+## Special Patterns
```
Curious here, more than sure. Most of the sub-points here have a single bullet point, bar one. Should we use paragraphs there instead?

benwuNYC replied:
I followed the [official .gitignore documentation ](https://git-scm.com/docs/gitignore/2.20.0) structure of listing the pattern formats. It may be easier to find quickly than if the information would be in paragraph form, but would be happy to fix it up to look more aesthetically pleasing.

adityasaky reviewed (changes requested):
Looking better and better, @itsbenwu! I think we're quite close now.

benwuNYC commented:
Thanks again for the re-review @adityasaky! Here's what I changed based on your recommendations

- Created an "Introduction section" for background information about `exclude` that describes where it originates and reworded and rewrote some definitions.
- Revised "Pattern Format" section to be solely based on `.gitignore` [documentation](https://git-scm.com/docs/gitignore), using in-line examples under recommendation of @lukpueh, taken from official `.gitignore` documentation.
- Removed the "Pattern Formats Example" section and incorporated some examples in line in the "Pattern Format" section. 
- Organized "Special Pattern" section to only include [documentation](https://github.com/cpburnz/python-path-specification/blob/master/pathspec/patterns/gitwildmatch.py#L59-L61) from `pathspec`.
- Reworded some sentences and fixed grammar, redone simple formatting issues with spacing/backticks, and organized paragraph and sentence structures. 

Here's what I think I have slight work to do:

- Try to further connect the "Special Patterns" and "Pattern Format" sections, provide further documentation (if needed), and try to remove any redundant/unnecessary information. 

Let me know what you think of these changes and if I should revert/edit anything further 😋

adityasaky commented on doc/source/exclude_patterns.md:
````diff
@@ -0,0 +1,45 @@
+
+The optional argument `exclude_patterns` in the `in_toto_run` API, also used by `--exclude` in the ```in-toto-run``` command line tool, applies [pathspec](http://python-path-specification.readthedocs.io) to compile [gitignore-style patterns](https://git-scm.com/docs/gitignore).
+
+Passed in arguments such as the paths and hashes of 'materials' (files before command execution) and 'products' (files after command execution) are not recorded with other information when generating link metadata files. If a pattern matches a directory, all files and subdirectories within that directory are also excluded.
````
```suggestion
Artifacts (materials and products) matched by an exclude pattern are not recorded when generating link metadata. If a pattern matches a directory, all files and subdirectories are also excluded.
```

benwuNYC replied:
Should we provide definitions of materials/products as previously, or is it assumed that users of the feature already know what materials and products are?

adityasaky replied:
I personally think it's safe to say they know what materials and products are if they're using exclude patterns.

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,45 @@
+  - Matched patterns will be unmatched. Escape the exclamation mark with a back-slash to match a literal exclamation mark (i.e., `\!`).
+
+## Source
+The source code for `pathspec` is available on [GitHub.](https://github.com/cpburnz/python-path-specification)
```
```suggestion
The source code for `pathspec` is available on [GitHub](https://github.com/cpburnz/python-path-specification).
```

benwuNYC replied:
Fixed :)

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,45 @@
+  - A pattern ending with a slash will match all descendant paths if it is a directory level but not if it is a regular file. This is equivalent to `{pattern}/**`.
+
+### Pattern that is a slash 
+  - According to 'git check-ignore' (v2.4.1), a single '/' does not match any file
```
```suggestion
  - According to 'git check-ignore' (v2.4.1), a single `/` does not match any file
```

adityasaky reviewed (commented):
Couple more minor nits.

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,45 @@
+  - Serves as a comment (neither includes or excludes files). Escape the hash with a back-slash to match a literal hash (i.e., `\#`). 
+
+### Pattern beginning with exclamation mark 
+  - Matched patterns will be unmatched. Escape the exclamation mark with a back-slash to match a literal exclamation mark (i.e., `\!`).
```
```suggestion
  - This has the effect of negating the rest of the pattern. Artifacts matched by the rest of the pattern will be the only ones recorded, and everything else will be excluded. Escape the exclamation mark with a back-slash to match a literal exclamation mark (i.e., `\!`).
```

adityasaky reviewed (approved):
I think this looks quite good to me, @itsbenwu! Thanks again for picking this up.

Can you clean up commits / sign off for DCO?

Also, we should add links to this document in the relevant docstrings.

adityasaky commented on doc/source/exclude_patterns.md:
````diff
@@ -0,0 +1,38 @@
+
+## Documentation
+
+Documentation for ```pathspec``` is available on [Read the Docs](http://python-path-specification.readthedocs.io/).   
````
```suggestion
- [```pathspec```](http://python-path-specification.readthedocs.io/)
```

Perhaps?

adityasaky replied:
and same for gitignore

benwuNYC committed:
Added exclude-pattern to index.rst

Signed-off-by: Benjamin Wu <benjaminwu1337@gmail.com>

adityasaky commented on doc/source/index.rst:
```diff
@@ -15,4 +15,5 @@ found at `in-toto.io <https://in-toto.io>`_.
    api
    model
    layout-creation-example
+   exclude-patterns <Will provide link when available>
```
We don't need a link unless it's an external document. Take a look at the others above vs the demo which is an external link. :)

adityasaky commented on doc/source/exclude_patterns.md:
```diff
@@ -0,0 +1,35 @@
+# Exclude Pattern Settings 
```
I guess this file came back during a rebase?

benwuNYC replied:
Oops! Fill fix it up and squash the commits

benwuNYC committed:
Update exclude-patterns.md

Signed-off-by: Benjamin Wu <benjaminwu1337@gmail.com>

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,33 @@
+
+## Pattern Formats
+  - A literal hash serves as a comment. Escape the hash with a back-slash to match a literal hash (i.e., `\#`). 
+  - A exclamation mark negates the rest of the pattern. Artifacts matched by the rest of the pattern will be the only ones recorded, and everything else will be excluded. Escape the exclamation mark with a back-slash to match a literal exclamation mark (i.e., `\!`).
```
I don't think this is true. Please double check with the code that the documentation you write is accurate.

lukpueh replied:
Generally, `#` and `!` special characters don't really seem useful for the purpose of exclude patterns. I'd probably mention them further down in the list and emphasize on the fact that they need to be escaped if they should be matched as literals.

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,33 @@
+## Pattern Formats
+  - A literal hash serves as a comment. Escape the hash with a back-slash to match a literal hash (i.e., `\#`). 
+  - A exclamation mark negates the rest of the pattern. Artifacts matched by the rest of the pattern will be the only ones recorded, and everything else will be excluded. Escape the exclamation mark with a back-slash to match a literal exclamation mark (i.e., `\!`).
+  - The slash is a directory seperator. Separators at beginning or middle (or both) of a pattern are relative to directory level of particular .gitignore file. Separators at the end of the pattern only  match directories. (e.g., `doc/frotz/` matches `doc/frotz` directory but not `a/doc/frotz` directory; however `frotz/` matches `frotz` and `a/frotz`).
```
Why do you mention the `.gitignore` file here? Also, this paragraph misses a few articles, and has a typo (seperator).

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,33 @@
+
+## Special Cases
+### Pattern beginning with a slash
+  - Match paths directly on the root directory instead of descendant paths. Escape the relative root conversion with an forward slash to match absolute paths. (i.e., 
```
This is the only imperative-style line, all others are descriptive-style. Please make consistent!

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,33 @@
+  - A literal hash serves as a comment. Escape the hash with a back-slash to match a literal hash (i.e., `\#`). 
+  - A exclamation mark negates the rest of the pattern. Artifacts matched by the rest of the pattern will be the only ones recorded, and everything else will be excluded. Escape the exclamation mark with a back-slash to match a literal exclamation mark (i.e., `\!`).
+  - The slash is a directory seperator. Separators at beginning or middle (or both) of a pattern are relative to directory level of particular .gitignore file. Separators at the end of the pattern only  match directories. (e.g., `doc/frotz/` matches `doc/frotz` directory but not `a/doc/frotz` directory; however `frotz/` matches `frotz` and `a/frotz`).
+  - A single `/` does not match any file.
```
Maybe you can move up the information from below ("Pattern beginning with a slash") here. It feels like the fit together. :)

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,33 @@
+  - A exclamation mark negates the rest of the pattern. Artifacts matched by the rest of the pattern will be the only ones recorded, and everything else will be excluded. Escape the exclamation mark with a back-slash to match a literal exclamation mark (i.e., `\!`).
+  - The slash is a directory seperator. Separators at beginning or middle (or both) of a pattern are relative to directory level of particular .gitignore file. Separators at the end of the pattern only  match directories. (e.g., `doc/frotz/` matches `doc/frotz` directory but not `a/doc/frotz` directory; however `frotz/` matches `frotz` and `a/frotz`).
+  - A single `/` does not match any file.
+  - Leading double asterisks match any proceeding path segments (e.g., `**/foo` matches file or directory `foo` and `**/foo/bar` matches file or directory `bar` anywhere that is directly under directory `foo`).
```
Do you mean "preceding"?

lukpueh replied:
Minor nit: I think this makes it a little less ambiguous.
```suggestion
  - Leading double asterisks match any proceeding path segments (e.g., `**/foo` matches file or directory `foo` anywhere, and `**/foo/bar` matches file or directory `bar` anywhere that is directly under directory `foo`).
```

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,33 @@
+  - The slash is a directory seperator. Separators at beginning or middle (or both) of a pattern are relative to directory level of particular .gitignore file. Separators at the end of the pattern only  match directories. (e.g., `doc/frotz/` matches `doc/frotz` directory but not `a/doc/frotz` directory; however `frotz/` matches `frotz` and `a/frotz`).
+  - A single `/` does not match any file.
+  - Leading double asterisks match any proceeding path segments (e.g., `**/foo` matches file or directory `foo` and `**/foo/bar` matches file or directory `bar` anywhere that is directly under directory `foo`).
+  - Trailing double asterisks matches any preceeding path segments (e.g., `abc/**` matches all files inside directory `abc`). 
```
```suggestion
  - Trailing double asterisks match any preceding path segments (e.g., `abc/**` matches all files inside directory `abc`). 
```

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,33 @@
+  
+  More information about pattern formats found in [`gitignore`](https://git-scm.com/docs/gitignore). 
+
+## Special Cases
```
What is the difference between special cases and pattern formats?

adityasaky replied:
This was my idea, I wondered if it was handy to have a section for pattern formats, and another for corner cases like the one that triggered the original issue.

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,33 @@
+  - Trailing double asterisks matches any preceeding path segments (e.g., `abc/**` matches all files inside directory `abc`). 
+  - A slash followed by two consecutive asterisks and a slash matches path segments between two directories (e.g., `a/**/b` matches `a/b`, `a/x/b`, `a/x/y/b` and so on). 
+  - Single asterisks match everything except a slash. 
+  - Question marks match any one character except a slash.
```
IMO, these two should be on top in this list.

lukpueh replied:
And the range characters too `[...]`.

lukpueh reviewed (changes requested):
Nice document, @itsbenwu! 🎉   I left a few comments inline. Please address and we should soon be able to merge.

adityasaky committed:
Minor fixes:

- Remove duplicate file
- Fix link in index.rst
- Reorganize some points
- Add some clarifying examples

Signed-off-by: Aditya Sirish <aditya@saky.in>

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,46 @@
+  segments between two directories (e.g., `a/**/b` matches `a/b`, `a/x/b`,
+  `a/x/y/b` and so on).
+- A pattern ending with a slash will match all descendant paths. This is
+  equivalent to `{pattern}/**`.
```
Maybe we can put:
- If a pattern matches a directory, all files and subdirectories are also excluded.
- Trailing double asterisks match any succeeding path segments (e.g., `abc/**`
matches all files inside directory `abc`).
- A pattern ending with a slash will match all descendant paths

all in one bullet point?

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,46 @@
+  or middle (or both) of a pattern are relative to the current directory.
+  Separators at the end of the pattern only match directories. (e.g.,
+  `doc/frotz/` matches `doc/frotz` directory but not `a/doc/frotz` directory;
+  however `frotz/` matches `frotz` and `a/frotz`).
```
I don't understand how these examples relate to the rule "Separators at the end of the pattern only match directories".

And I don't understand why `frotz/` matches `a/frotz`. Wouldn't we need a pattern `**/frotz` for for that instead?

adityasaky replied:
```
>>> from pathspec import PathSpec 
>>> from in_toto.runlib import _apply_exclude_patterns 
>>> exclude_patterns = ["frotz/"] 
>>> exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_patterns)
>>> _apply_exclude_patterns(["frotz/", "a/frotz/"], exclude_filter)
[]
>>> _apply_exclude_patterns(["frotz/", "a/frotz"], exclude_filter)
['a/frotz']
```

I've updated the document to read `a/frotz/` instead.

lukpueh reviewed (commented):
Excellent clean-up, @adityasaky!

I have one confusion and one suggestion inline.

lukpueh commented on doc/source/exclude-patterns.md:
```diff
@@ -0,0 +1,46 @@
+  `[a-z]`).
+- A forward slash indicates a directory separator.
+- A single `/` does not match any file.
+- Separators at the beginning or middle (or both) of a pattern are relative to
+  the current directory. Separators at the end of the pattern only match
+  directories. (e.g., `doc/frotz/` matches the `doc/frotz/` directory but not
+  `a/doc/frotz/`; however `frotz/` matches `frotz/` and `a/frotz/`).
```
> Separators at the beginning or middle (or both) of a pattern are relative to  the current directory.

Can we add an example with a separator at the beginning, and middle (and both) of a pattern?

>  Separators at the end of the pattern only match directories. (e.g., `doc/frotz/` matches the `doc/frotz/` directory but not `a/doc/frotz/`, however `frotz/` matches `frotz/` and `a/frotz/`.

I still don't understand the choice of the examples here. The rule doesn't really explain why  `doc/frotz/` doesn't match `a/doc/frotz/`. Also, "`doc/frotz/` matches `doc/frotz/`" seems redundant with "`frotz/` matches `frotz/`".

adityasaky replied:
> The rule doesn't really explain why doc/frotz/ doesn't match a/doc/frotz/.

That's a fair point, and it seems to be by design. Perhaps worth looking up the design decision by git / pathspec.

> Also, "doc/frotz/ matches doc/frotz/" seems redundant with "frotz/ matches frotz/".

No, actually, I think it's worth showing this because of the following:

```
>>> exclude_patterns = ["doc/frotz/"]
>>> exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_patterns)
>>> _apply_exclude_patterns(["a/doc/frotz/", "doc/frotz/"], exclude_filter)
['a/doc/frotz/']
```

```
>>> exclude_patterns = ["frotz/"]
>>> exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_patterns)
>>> _apply_exclude_patterns(["a/doc/frotz/", "doc/frotz/"], exclude_filter)
[]
```

adityasaky replied:
> Can we add an example with a separator at the beginning, and middle (and both) of a pattern?

I think `doc/frotz` highlights that actually, and that's why it's not redundant with `frotz/`. I'll add something for the beginning.

```
>>> exclude_patterns = ["/frotz/"]
>>> exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_patterns)
>>> _apply_exclude_patterns(["a/doc/frotz/", "doc/frotz/"], exclude_filter)
['a/doc/frotz/', 'doc/frotz/']
>>> _apply_exclude_patterns(["a/doc/frotz/", "frotz/"], exclude_filter)
['a/doc/frotz/']
```

lukpueh replied:
Thanks for the clarifications! Just approved the PR. I think you need to approve it too. :)

SantiagoTorres commented:
@lukpueh I think you're scheduled to review this. I personally think it LGTM. Let me know if you need me to eye this out and merge.

lukpueh reviewed (approved):
Excellent work, @adityasaky! Thanks. :)

adityasaky reviewed (approved):
Thanks again for working on this, @itsbenwu! And, uh, LGTM!

shibumi commented:
Hey @adityasaky and @itsbenwu  do we have a set of specific test cases already? I would like to test my go-pathspec fork if it satisfies all of our requirements.

lukpueh commented:
@shibumi, this is what we have:

https://github.com/in-toto/in-toto/blob/45bbab858f356a986fd44a916b99d329283f6804/tests/test_runlib.py#L45-L88

Would be nice to compare to the newly added docs and see if we should add more tests. :)

shibumi commented:
@lukpueh What I see missing on the first sight:

* `**`-Operator
* `foo` should match on `/bardir/foodir/foo`

adityasaky commented:
Good call to expand the tests using the new docs, @lukpueh & @shibumi. I wonder if it's something a new student can take up, if we ticketize it?

shibumi commented:
@adityasaky adding a few test cases should be doable by a new student :) Let me create a ticket for it in in-toto (there is already one for the Go implementation).

shibumi commented:
@adityasaky here you go: https://github.com/in-toto/in-toto/issues/391 :)
